src/include/type/value.h

```diff
@@ -0,0 +1,266 @@
+    TERRIER_ASSERT(type_id_ == TypeId::INTEGER, "The type must be a integer");
+    return &value_.integer;
+  }
+
+  /**
+   * Get contents of a bigint value
+   * @return int64_t
+   */
+  const int64_t *GetBigIntValue() const {
+    TERRIER_ASSERT(type_id_ == TypeId::BIGINT, "The type must be a bigint");
+    return &value_.bigint;
+  }
+
+  /**
+   * Get contents of a decimal value
+   * @return double
+   */
+  const double *GetDecimalValue() const {
+    TERRIER_ASSERT(type_id_ == TypeId::DECIMAL, "The type must be a decimal");
+    return &value_.decimal;
+  }
+
+  /**
+   * Get contents of a timestamp value
+   * @return timestamp_t
+   */
+  const timestamp_t *GetTimestampValue() const {
+    TERRIER_ASSERT(type_id_ == TypeId::TIMESTAMP, "The type must be a timestamp");
+    return &value_.timestamp;
+  }
+
+  /**
+   * Get contents of a date value
+   * @return date_t
+   */
+  const date_t *GetDateValue() const {
+    TERRIER_ASSERT(type_id_ == TypeId::DATE, "The type must be a date");
+    return &value_.date;
+  }
+
+  /**
+   * Compare values for equality
+   * @param rhs
+   * @return true if equal, false if not.
+   */
+  bool operator==(const Value &rhs) const;
+
+  /**
+   * Hash the value
+   * @return hashed value
+   */
+  hash_t Hash() const {
+    hash_t ret_hash = 0;
+
+    switch (type_id_) {
+      case TypeId::BOOLEAN:
+        ret_hash = HashUtil::Hash(GetBooleanValue());
+        break;
+
+      case TypeId::TINYINT:
+        ret_hash = HashUtil::Hash(GetTinyIntValue());
+        break;
+
+      case TypeId::SMALLINT:
+        ret_hash = HashUtil::Hash(GetSmallIntValue());
+        break;
+
+      case TypeId::INTEGER:
+        ret_hash = HashUtil::Hash(GetIntValue());
+        break;
+
+      case TypeId::BIGINT:
+        ret_hash = HashUtil::Hash(GetBigIntValue());
+        break;
+
+      case TypeId::DATE:
+        ret_hash = HashUtil::Hash(GetDateValue());
+        break;
+
+      case TypeId::DECIMAL:
+        ret_hash = HashUtil::Hash(GetDecimalValue());
+        break;
+
+      case TypeId::TIMESTAMP:
+        ret_hash = HashUtil::Hash(GetTimestampValue());
+        break;
+
+      default:
+        TERRIER_ASSERT(false, "unsupported type");
+        break;
+    }
+    return ret_hash;
+  }
+
+ protected:
+  /**
+   * Holds the type and value
+   */
+  union Val {
+    /**
+     * booleans
+     */
+    boolean_t boolean;
+    /**
+     * tiny integers
+     */
+    int8_t tinyint;
+    /**
+     * small integers
+     */
+    int16_t smallint;
+    /**
+     * Integers
+     */
+    int32_t integer;
+    /**
+     * Big integers
+     */
+    int64_t bigint;
+    /**
+     * Decimals, i.e. floating point values
+     */
+    double decimal;
+    /**
+     * Timestamps
+     */
+    timestamp_t timestamp;
+    /**
+     * Date values
+     */
+    date_t date;
+    /**
+     * variable length strings, without a null terminator. Length specified by var_len
+     */
+    char *varchar;
```

Comment:
We prefer byte * over char * in the storage layer, but maybe it makes sense at this layer since we're actually reasoning about a VARCHAR at this point?